### DISCUSSION TITLE
Does this exporter support IBM DB2 and Oracle DB?

### DISCUSSION THREAD

**Discussion by shingou0120** [burningalchemist/sql_exporter#299] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I am looking for a solution to export metrics from DB2 and Oracle database.
Current we use https://github.com/albertodonato/query-exporter as our biz monitor solution.
However, the configuration format in albertodonato/query-exporter is hard to implement to helm chart.

**Describe the solution you'd like**
The solution I need is to make db, metrics and query sql more flexible and reusable with helm chart
There are many sites in our company such as A site, B site and C site, and we should export the metrics with the same SQL from DB of each site.
For example, 
metric: `goods_inventory_count`
query: `select count from tbl_inventory where <condition>`
Both of those above are the same, and I should deploy 3 sql_exporter in 3 sites to connect different DB.
So, I want to build a helm chart package to form the same metrics and queries.

After I survey overall, sql_exporter seems to fit my needs except that DB2 and oracle are not supported. :(

**Describe alternatives you've considered**
Currently not. Only if I try to build the config template for query-exporter, and I am sure it's really not easy.

If the DB2 and Oracle could be supported, it's so appreciated, thanks a lot.

**Comment by burningalchemist** [comment, marked as answer]
Hi @shingou0120, I'm sorry for the late response - had many things around.

Generally, any database that has a driver in golang, supporting `sql/database` interface, and supported by `dburl` can be used by `sql_exporter`.

Following your case I found next things:

- Oracle DB is supported. :+1: The suitable driver for Oracle DB is likely https://github.com/sijms/go-ora.

- As for IBM DB2 - it's a bit tricky. There is a driver written in go, but it's not supported by `dburl` dependency, so a pull request is needed on the [upstream](https://github.com/xo/dburl).

These drivers are rather specific, and aren't available out of the box, but it's not difficult to add them (one of them at the moment):

1. add a driver to one of the maps [here](https://github.com/burningalchemist/sql_exporter/blob/28624dbbb6e1abf278684db530f8532e26806987/drivers_gen.go#L17)
2. generate `drivers.go` by running `make drivers-all` command
3. once `drivers.go` is generated, you can compile the binary with `make build`
4. done.

P.S. You can also create the binary only with custom drivers - populate the list `custom`, and run `make drivers-custom` before compiling the binary.

Let me know if you have any questions, I'm happy to assist. 👍

**Comment by JustNZ** [comment]
Hey @burningalchemist,
I added the driver you mentioned to `custom` section, did the `make` for `drivers-custom` and after that the `build`.

After installing the binary and adding `oracle://...` to the connection string I get the following error on execution of the binary:
`sql: unknown driver \"oracle\" (forgotten import?)"`

Do you know what I did wrong?
Is there any way for you to add the mentioned oracle driver directly to this project?

**Comment by burningalchemist** [comment]
Hey @JustNZ,

I just tested it, and it works fine. It's also that you need to use `github.com/sijms/go-ora/v2`. This might be the cause of your issue.

Here's a short screencast, hope it helps:

https://github.com/burningalchemist/sql_exporter/assets/18210256/95883be5-3811-43bd-96b9-21250fed9dea

As for adding `oracle` driver by default - it's unlikely, since there's no way to test it, it also increases the application footprint, which is against the idea of this app - to be lightweight and simple. It's also a pretty niche solution. I'm also considering moving previously added `Vertica` and `Snowflake` drivers off.

I think adding a simple guide to add your own driver is the best option as it gives control and flexibility.

**Comment by JustNZ** [comment]
Hey @burningalchemist ,

yes seems like the issue was by not using the v2 version of the oracle driver.
Thanks a lot for the lightning fast response and it is working like a charm now.

And keeping this project lightweight does make sense and if there is the possibility to add own drivers then it is more than perfect.

Thanks!
